opendrift/models/oceandrift2.py

```diff
@@ -0,0 +1,791 @@
+                w = self.elements.terminal_velocity
+
+                # Diffusivity and its gradient at z
+                zi = np.round(z_index(-self.elements.z)).astype(np.uint16)
+                Kz = Kprofiles[zi, range(Kprofiles.shape[1])]
+                dKdz = gradK[zi, range(Kprofiles.shape[1])]
+
+                # Visser et al. 1997 random walk mixing
+                # requires an inner loop time step dt such that
+                # dt << (d2K/dz2)^-1, e.g. typically dt << 15min
+                #
+                # NB: In the last term Kz is evaluated in zi, while
+                # it should be evaluated in (self.elements.z - dKdz*dt_mix)
+                # This is not expected have large impact on the result
+                R = 2*np.random.random(self.num_elements_active()) - 1
+                r = 1.0/3
+                # New position  =  old position   - up_K_flux   + random walk
+                self.elements.z = self.elements.z - self.elements.moving*(
+                    dKdz*dt_mix - R*np.sqrt((Kz*dt_mix*2/r)))
+
+                # Reflect from surface
+                reflect = np.where(self.elements.z >= 0)
+                if len(reflect[0]) > 0:
+                    self.elements.z[reflect] = -self.elements.z[reflect]
+
+                # Reflect elements going below seafloor
+                bottom = np.where(np.logical_and(self.elements.z < Zmin, self.elements.moving == 1))
+                if len(bottom[0]) > 0:
+                    logger.debug('%s elements penetrated seafloor, lifting up' % len(bottom[0]))
+                    self.elements.z[bottom] = 2*Zmin[bottom] - self.elements.z[bottom]
+
+                # Advect due to buoyancy
+                self.elements.z = self.elements.z + w*dt_mix*self.elements.moving
+
+                # Put the particles that belonged to the surface slick
+                # (if present) back to the surface
+                self.elements.z[surface] = 0.
+
+                # Formation of slick and wave mixing for surfaced particles
+                # if implemented for this class
+                self.surface_stick()
+                self.surface_wave_mixing(dt_mix)
+
+                # Let particles stick to bottom
+                bottom = np.where(self.elements.z < Zmin)
+                if len(bottom[0]) > 0:
+                    logger.debug('%s elements reached seafloor, set to bottom' % len(bottom[0]))
+                    self.interact_with_seafloor()
+                    self.bottom_interaction(Zmin)
+
+                if store_depths is not False:
+                    depths[i, :] = self.elements.z
+
+            self.timer_end('main loop:updating elements:vertical mixing')
+
+            if store_depths is not False:
+                return depths
+            else:
+                return None
+
+        def animate_vertical_distribution(self, depths=None, maxdepth=None, bins=50, filename=None, subsamplingstep=1):
```

Comment:
methods like this I guess should probably not belong to Simulation class, but to a Result class. Though we decided to not have a Result class, so this should then rather be moved to the separate analysis package(?)

Reply:
Yes. It has to work on the results.